codec-http2/src/main/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributor.java

```diff
@@ -0,0 +1,328 @@
+                        pState.offerAndInitializePseudoTime(state);
+                        pState.isActiveCountChangeForTree(state.isActiveCountForTree);
+                    }
+                }
+            }
+
+            @Override
+            public void onPriorityTreeParentChanging(Http2Stream stream, Http2Stream newParent) {
+                Http2Stream parent = stream.parent();
+                if (parent != null) {
+                    State state = state(stream);
+                    if (state.isActiveCountForTree != 0) {
+                        State pState = state(parent);
+                        pState.remove(state);
+                        pState.isActiveCountChangeForTree(-state.isActiveCountForTree);
+                    }
+                }
+            }
+        });
+    }
+
+    @Override
+    public void updateStreamableBytes(StreamState state) {
+        state(state.stream()).updateStreamableBytes(state.streamableBytes(),
+                                                    state.hasFrame() && state.isWriteAllowed());
+    }
+
+    @Override
+    public boolean distribute(int maxBytes, Writer writer) throws Http2Exception {
+        checkNotNull(writer, "writer");
+
+        // As long as there is some active frame we should write at least 1 time.
+        if (connectionState.isActiveCountForTree == 0) {
+            return false;
+        }
+
+        // The goal is to write until we write all the allocated bytes or are no longer making progress.
+        // We still attempt to write even after the number of allocated bytes has been exhausted to allow empty frames
+        // to be sent. Making progress means the active streams rooted at the connection stream has changed.
+        int oldIsActiveCountForTree;
+        do {
+            oldIsActiveCountForTree = connectionState.isActiveCountForTree;
+            // connectionState will never be active, so go right to its children.
+            maxBytes -= distributeToChildren(maxBytes, writer, connectionState);
+        } while (connectionState.isActiveCountForTree != 0 &&
+                (maxBytes > 0 || oldIsActiveCountForTree != connectionState.isActiveCountForTree));
+
+        return connectionState.isActiveCountForTree != 0;
+    }
+
+    /**
+     * Sets the amount of bytes that will be allocated to each stream. Defaults to 1KiB.
+     * @param minAllocationChunk the amount of bytes that will be allocated to each stream. Must be > 0.
+     */
+    public void allocationQuantum(int allocationQuantum) {
+        if (allocationQuantum <= 0) {
+            throw new IllegalArgumentException("allocationQuantum must be > 0");
+        }
+        this.allocationQuantum = allocationQuantum;
+    }
+
+    private int distribute(int maxBytes, Writer writer, State state) throws Http2Exception {
+        if (state.isActive) {
+            int nsent = min(maxBytes, state.streamableBytes);
+            state.write(nsent, writer);
```

Comment:
I think I recall discussing this...We don't want to write headers until we are able to write data too. sgtm.